src/key_store/mod.rs

```diff
@@ -0,0 +1,222 @@
+        // We unwrap here, because the two functions claiming write locks (this
+        // one and `generate_key_package`) only hold the lock very briefly and
+        // should not panic during that period.
+        let mut kpbs = self.init_key_package_bundles.write().unwrap();
+        kpbs.remove(kp_hash)
+    }
+
+    /// Retrieve a `CBGuard` from the key store given the `SignaturePublicKey`
+    /// of the corresponding `Credential`. The `CBGuard` can be dereferenced to
+    /// obtain an `RwLock` on the desired `CredentialBundle`. Returns an error
+    /// if no `CredentialBundle` can be found corresponding to the given
+    /// `SignaturePublicKey`.
+    pub(crate) fn get_credential_bundle<'key_store>(
+        &'key_store self,
+        signature_public_key: &'key_store SignaturePublicKey,
+    ) -> Option<CBGuard> {
+        let cbs = self.credential_bundles.read().unwrap();
+        if !cbs.contains_key(signature_public_key) {
+            return None;
+        }
+        Some(CBGuard {
+            cbs,
+            index: signature_public_key,
+        })
+    }
+
+    /// Generate a fresh `KeyPackageBundle` with the given parameters, store it
+    /// in the `KeyStore` and return the corresponding `KeyPackage`. Throws an
+    /// error if no `CredentialBundle` can be found in the `KeyStore`
+    /// corresponding to the given `Credential` or if an error occurred during
+    /// the creation of the `KeyPackageBundle`.
+    pub fn generate_key_package(
+        &self,
+        ciphersuites: &[CiphersuiteName],
+        credential: &Credential,
+        extensions: Vec<Box<dyn Extension>>,
+    ) -> Result<KeyPackage, KeyStoreError> {
+        let credential_bundle = self
+            .get_credential_bundle(credential.signature_key())
+            .ok_or(KeyStoreError::NoMatchingCredentialBundle)?;
+        let kpb = KeyPackageBundle::new(ciphersuites, &credential_bundle, extensions)?;
+        let kp_hash = kpb.key_package().hash();
+        // We unwrap here, because the two functions claiming write locks (this
+        // one and `take_key_package_bundle`) only hold the lock very briefly
+        // and should not panic during that period.
+        let mut kpbs = self.init_key_package_bundles.write().unwrap();
+        kpbs.insert(kp_hash.clone(), kpb);
+        let kp = kpbs.get(&kp_hash).unwrap().key_package().clone();
+        Ok(kp)
+    }
+
+    /// Generate a fresh `CredentialBundle` with the given parameters and store
+    /// it in the `KeyStore`. Returns the corresponding `Credential` or an error
+    /// if the creation of the `CredentialBundle` fails.
+    pub fn generate_credential(
+        &self,
+        identity: Vec<u8>,
+        credential_type: CredentialType,
+        signature_scheme: SignatureScheme,
+    ) -> Result<Credential, KeyStoreError> {
+        let cb = CredentialBundle::new(identity, credential_type, signature_scheme)?;
+        let signature_key = cb.credential().signature_key().clone();
```

Comment:
Why all this cloning? Appears to be a little excessive.

Reply:
I believe we have to clone here at least once, because right afterwards, the `CredentialBundle` is consumed by `insert`. I've refactored a bit to get rid of a clone. Both `generate_` functions currently return a full `Credential`/`KeyPackage` instance. We could also try returning a reference, but that would require another `Guard` struct and it would keep a read lock on the corresponding `HashMap`. What do you think? I think a clone is ok here.